### ISSUE TITLE
Not many Compiler issues tagged Help Wanted

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/roslyn#32441]
I've recently started contributing to Roslyn, and I am eager to contribute more.

I'm more interested in contributing to the compiler than to the IDE or to Analyzers. However there are not many issues tagged both with `Area-Compilers` and `help wanted`. Of those that are, I've gone through a few, and opened some pull requests. However less than 20 issues from the last year with both these tags are open, and those that are suitable/interest me are running a bit thin.

So a few requests here:

a) Could more compiler issues be marked `help wanted` in the future?

b) Similarly could open compiler issues be looked at and tagged `help wanted` or could someone point me towards issues that might be relevant?

c) I've noticed that there are no compiler issues marked both with `New Language Feature` and `help wanted`. Is there a reason for that?

d) If anyone does open an issue, which they think I might be able to help with, please do tag me in a comment!

I love C# and I really appreciate the work all of you put in to make such a great language and tooling experience! I'm looking to expand my knowledge of language and compiler design, and I feel like contributing to Roslyn is the perfect way to do so and give back to the community.

Thank you very much!

**Comment by YairHalberstadt** [comment]
@jaredpar 
CyrusNajmabadi suggested you would be the right person to talk to, so I'm tagging you in here.
Thanks

**Comment by jmarolf** [comment]
> a) Could more compiler issues be marked help wanted in the future?

There are [187](https://github.com/dotnet/roslyn/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3AArea-Compilers+label%3A%22help+wanted%22+) issues with `Area-Compiler`+`help-wanted` at this time.  How many do you want :)

> b) Similarly could open compiler issues be looked at and tagged help wanted or could someone point me towards issues that might be relevant?

I happily volunteer @jaredpar 	to do that.

> c) I've noticed that there are no compiler issues marked both with New Language Feature and help wanted. Is there a reason for that?

Yes, the implementation of language features by members of the community should start over on [csharplang](https://github.com/dotnet/csharplang/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc). Once the team approves the design of a feature (as per the [design-process](https://github.com/dotnet/csharplang#design-process)) then talk can begin about who has time to do it.  In general language design is not tracked in this repo.

**Comment by YairHalberstadt** [comment]
> There are 187 issues with Area-Compiler+help-wanted at this time. How many do you want :)

188 :)

There's only 18 open issues from the last year, 3 of which I have an open pull request against.

I'm a bit nervous to do older ones, in case they're no longer relevant or wanted. But maybe my my fears are unwarranted.

> Yes, the implementation of language features by members of the community should start over on csharplang. Once the team approves the design of a feature (as per the design-process) then talk can begin about who has time to do it. In general language design is not tracked in this repo.

And the team isn't looking for help with any specific parts of the implementation? There are hundreds of issues relating to NRTs. If there any reason none of them could be done by community members?

Thanks for your feedback!

**Comment by jmarolf** [comment]
@YairHalberstadt 	To be clear I am absolutely thrilled to hear that someone is interested in fixing compiler bugs. I also hear the complaint of triage not being up-to-date in some areas. I'll see what can be done.  I agree that some of [these](https://github.com/dotnet/roslyn/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22New+Language+Feature+-+Nullable+Reference+Types%22+label%3AArea-Compilers) could be fixed by a community member.

**Comment by jaredpar** [comment]
@YairHalberstadt 

> Could more compiler issues be marked help wanted in the future?

Yes certainly. Although for language features there tend to be a smaller window when "help wanted" features exist. Generally between when a feature first makes it into the master branch and about the fist update after a feature RTMs. During that time we usually make it through most of the easier issues / ones that need less context. After the first update generally the issues get fairly complex and require some deeper compat investigations before we can submit a fix. That contributes to why we generally have less "help wanted" issues than IDE (they're not nearly as compat constrained). 

> Similarly could open compiler issues be looked at and tagged help wanted or could someone point me towards issues that might be relevant?

Yes. I generally do a good portion of our triaging. I will be more mindful at marking appropriate issues as "help wanted" in the future.

> I've noticed that there are no compiler issues marked both with New Language Feature and help wanted. Is there a reason for that?

The compiler does the majority of feature work in separate branches. Features don't get merged back into master until they're fairly near completion. This is done to minimize noise on other teams. Pretty much as soon as a feature is merged into master it is used in other repositories around Microsoft. Merging half finished features to master means we're more likely to have significant churn later and hence cause issues for those repositories that bet on the first implementation. 

While in a separate branch we don't use issues to track bugs. Instead we have a light weight commenting style to track debt, bugs, etc ... A comment of the following form is used to track work:

``` csharp
// PROTOTYPE: need to get generics working in this code path
```

When merging back to master all unresolved PROTOTYPE comments are turned into issues. This is why you don't see much in terms of "help wanted" + New Language Feature. Those issues are all tracked as comments and most of the issues are resolved before merging.

Searching for PROTOTYPE comments in feature branches you're interested in though is likely to find more of the "help wanted" issues you're looking for.

> I feel like contributing to Roslyn is the perfect way to do so and give back to the community.

Your contributions have been much appreciated. Please keep them coming :smile:
